src/daemon/systemd-notify.c

```diff
@@ -0,0 +1,172 @@
+
+  if (connect(fd, &socket_addr.sa, offsetof(struct sockaddr_un, sun_path) + path_length) != 0)
+    return -errno;
+
+  ssize_t written = write(fd, message, message_length);
+  if (written != (ssize_t) message_length)
+    return written < 0 ? -errno : -EPROTO;
+
+  return 1; // Notified!
+}
+
+/* Notify the service manager that Netdata has finished startup successfully.
+ * This should be called only after we are sure we won’t exit due to
+ * some issue with the configuration or environment.
+ */
+int notify_ready(void) {
+  return notify("READY=1");
+}
+
+/* Notify the service manager that Netdata is reloading.
+ * This should be called at the start of a configuration reload, and if called notify_read() _MUST_ be called when the configuration reload finishes.
+ */
+int notify_reloading(void) {
+  /* A buffer with length sufficient to format the maximum UINT64 value. */
+  char reload_message[sizeof("RELOADING=1\nMONOTONIC_USEC=18446744073709551615")];
+  struct timespec ts;
+  uint64_t now;
+
+  /* Notify systemd that we are reloading, including a CLOCK_MONOTONIC timestamp in usec
+   * so that the program is compatible with a Type=notify-reload service. */
+
+  if (clock_gettime(CLOCK_MONOTONIC, &ts) < 0)
+    return -errno;
+
+  if (ts.tv_sec < 0 || ts.tv_nsec < 0 ||
+      (uint64_t) ts.tv_sec > (UINT64_MAX - (ts.tv_nsec / 1000ULL)) / 1000000ULL)
+    return -EINVAL;
+
+  now = (uint64_t) ts.tv_sec * 1000000ULL + (uint64_t) ts.tv_nsec / 1000ULL;
+
+  if (snprintf(reload_message, sizeof(reload_message), "RELOADING=1\nMONOTONIC_USEC=%" PRIu64, now) < 0)
+    return -EINVAL;
+
+  return notify(reload_message);
+}
+
+/* Request a service timeout extension from the service manager.
+ *
+ * The timeout should be a number in microseconds indicating the desired
+ * service timeout extension
+ */
+int notify_extend_timeout(uint64_t timeout) {
+  size_t msg_length = sizeof("EXTEND_TIMEOUT_USEC=18446744073709551615");
+  char message[msg_length];
+
+  snprintf(message, msg_length, "EXTEND_TIMEOUT_USEC=%lu", timeout);
+
+  return notify(message);
+}
+
+/* Notify the service manager that Netdata is stopping.
+ * This should be called during the clean exit path.
+ *
+ * The timeout should be a number in microseconds indicating the desired
+ * service timeout extension in the stop path (IOW, an upper bound on
+ * how long we think it will take to stop).
+ */
+int notify_stopping(uint64_t timeout) {
+  size_t msg_length = sizeof("STOPPING=1\nEXTEND_TIMEOUT_USEC=18446744073709551615");
+  /* A buffer with length sufficient to format the maximum UINT64 value. */
+  char stop_message[msg_length];
+
+  snprintf(stop_message, msg_length, "STOPPING=1\nEXTEND_TIMEOUT_USEC=%lu", timeout);
+
+  return notify(stop_message);
+}
+
+/* Send a status message to the service manager.
+ * These are used to indicate what step is happening during startup or shutdown.
+ */
+int notify_status(const char *message) {
+  size_t msg_length = strlen(message) + 7;
```

Comment:
Please add one additional character to the length to account for the null terminator.